add hybrid dev docs (#557)
okteto/docs: src/content/manifest/hybrid-development.mdx

### Problem to solve

Sometimes running the service locally is faster than the original approach taken in okteto where the service runs in the cloud and changes are synchronized from local to the remote development container. Another possibility can be if you’re using a framework that is IDE-heavy like on Java/Spring, IDEs tend to be pre-configured to run tests, start services, and connect debuggers. You lose some of that ‘click and you are done’ when you  run your development service in the cluster. This is where hybrid development comes in.

If you want to adopt this approach, you need to keep in mind that developing with the service running locally breaks the replicability between OSs in addition to the added latency when accessing the rest of the services in the cluster. 

### How hybrid development works?
Usually, okteto allows you to develop a service locally and run it in a production-like cloud environment synchronizing the local changes with the development container running in the cloud (sync mode) but okteto also allows you to use another approach in which the service to be developed runs locally and the rest of the application components run in the cloud (hybrid mode).

> 

Author: @rlamana should we include an image about the arch here? similar to https://github.com/okteto/okteto/blob/master/docs/okteto-architecture.png

Reviewer: If I create a diagram, I think we should also add a description of the how, to me this paragraph talks more about what it is rather that how it is achieved...

Reviewer: I found a diagram we already had something like this:

<img width="561" alt="Screenshot 2023-12-14 at 15 39 12" src="https://github.com/okteto/docs/assets/237819/3a990298-de59-4c82-a485-d3206ab0a5fe">

Reviewer 2: @rlamana could we also have the image representing the state before `okteto up`? and in this image, I would replace the `api` text inside the namespace by `SSH Proxy`

Reviewer: @pchico83 What about something like this:

<img width="1070" alt="Screenshot 2023-12-15 at 12 01 07" src="https://github.com/okteto/docs/assets/237819/1b5db928-2609-446c-bff7-55e54efa9bd0">

Author: I've added the last one ATM

Reviewer: @pchico83 can you review the diagram?

---
title: Hybrid development
description: Hybrid development allows you to develop and run a service locally while the other components of your app run in the cloud.
sidebar_label: Hybrid development
id: hybrid-development
---

import Image from '@theme/Image';

> Okteto offers two different approaches to service development: *sync* ot *hybrid* mode. The command affected by this is the [`okteto up`](/reference/cli/#up) command being configurable at manifest level.

Reviewer 3: ```suggestion
Okteto offers two different approaches to service development: *sync* ot *hybrid* mode. The command affected by this is the [`okteto up`](reference/cli.mdx#up) command being configurable at manifest level.
```

You need to link directly to the .mdx file, else this will break when we do versioning.

More details here: https://github.com/okteto/docs/pull/295

Author: great 041bf0170a82d66651c665dd74636951a932d0bc

The default mode used is the *sync* mode in which changes made to the local service are synchronized with the remote development container deployed in the cluster. On the other hand, the *hybrid* mode allows to run the development service locally and the rest of the application components in the cloud.

> ### Problem to solve

Reviewer 4: ```suggestion
### Deciding which mode to use
```

Maybe something like this? "Problem to solve" doesn't seem fitting for public docs since "problem" has a negative connotation. I propose something like this instead 😄

Author: 3b791f1eb3ff23bf057a17da1065f5533b18133a

> Sometimes running the service locally is faster than the original approach taken in okteto where the service runs in the cloud and changes are synchronized from local to the remote development container. Another possibility can be if you’re using a framework that is IDE-heavy like on Java/Spring, IDEs tend to be pre-configured to run tests, start services, and connect debuggers. You lose some of that ‘click and you are done’ when you  run your development service in the cluster. This is where hybrid development comes in.

Reviewer 4: ```suggestion
There are situations when running a service locally is faster than synchronizing code with a remote development container. For example: when doing frontend-only development or when working in an IDE-heavy context (e.g. Java/Spring) where the IDE may be pre-configured to run tests, start services, and connect debuggers. 

When using Okteto's `hybrid` mode you can maintain your local development experience for the specific service you're working on while still leveraging the benefits of hosting the rest of your application's components in the remote development container.
```

Wondering if something like this might work?

Author: af9a32b317861e294945db0b4e0dd006fdc2c9e8

> > Using *hybrid* development, you need to keep in mind that developing with the service running locally breaks the replicability between OSs in addition to the added latency when accessing the rest of the services in the cluster. 

Author: 1bbd05c534a2e3b502ffb109e5c180001b530e66

> ### How hybrid development works? 

Reviewer 4: ```suggestion
### How hybrid development works
```

> In the case of hybrid development, in order to allow communication from the cluster to the local machine and vice versa an SSH server is still deployed in the development container. So if we want to expose the local service in the development container we will use the [`reverse`](/reference/okteto-manifest.mdx/#reverse-string-optional) configuration for the development environment. In case we want to forward to the local machine a service exposed in the cluster, we can use the [`forward`](/reference/okteto-manifest.mdx/#forward-string-optional) configuration.

Reviewer 4: ```suggestion
To allow communication to and from the cluster and your local machine an SSH server is deployed in the remote development container. You can expose the local service to the development container using the [`reverse`](/reference/okteto-manifest.mdx/#reverse-string-optional) configuration for the development environment. If you want to forward a service exposed in the cluster to your local machine you can use the [`forward`](/reference/okteto-manifest.mdx/#forward-string-optional) configuration.
```

Author: d43a8d7150223e44aa96a02b8d7c57b5e45deb13

<p align="center">
  <Image
    src={require("@site/static/img/hybrid-development-arch.png").default}
    alt="Hybrid development architecture"
    width="850"
  />
</p>

> By using the *hybrid* mode, the process running locally is intended to have an environment similar to the one it would have if it were running in *sync* mode. As part of this, the following values are added to the local process running the service in *hybrid* mode as environment variables, in order of highest to lowest prevalence:

Reviewer 4: ```suggestion
When using `hybrid` mode the process running locally is designed to behave similarly as if it were running in `sync` mode. To achieve this the following environment variables are added to the local process running the service in `hybrid` mode. This list is ordered from highest to lowest prevalence:
```

Author: ce55671f47de5b99d0755d85353d8ac2a21101ff

> - Environment variables declared in the dev section of the manifest.

Reviewer 4: ```suggestion
- Environment variables declared in the `dev` section of the manifest
```

Author: a261c50d657ca80b711e92a88c878e5865bc1943

- Environment variables of the remote development container
- Environment variables of the configmap associated with the dev environment

> - User-defined secrets

Reviewer 4: ```suggestion
- User-defined variables
```

@pchico83 suggesting this change as part of making "secrets" -> "variables" language more consistent

- Environment variables declared in the image of the remote development container

> ### How to use the development in *hybrid* mode

Reviewer 4: ```suggestion
### How to develop in hybrid mode
```

> Using *hybrid* mode is very simple. If you want to use the *hybrid* mode, you only need to configure the necessary ports to expose the service running on your local machine to the development container in remote so that the rest of the components running on the cluster can communicate with the development container if needed as if it were the one running the application remotely. To do that, you can make use of the [reverse](reference/manifest/#reverse-string-optional) field, which will expose the indicated port of your local machine to the desired port on the development container.

Reviewer 4: ```suggestion
To enable `hybrid` mode you only need to configure the necessary ports to expose the local service to the remote development container. This allows the components running in the cluster to communicate with the local service as if it was running directly in the cluster. 

To do this you can make use of the [reverse](reference/manifest/#reverse-string-optional) field in your Okteto manifest, which will expose the configured port of your local machine to the desired port on the development container.
```

Author: 3eba66b255a3178d3016b3dda7a8576cc10a860c

```yaml
dev:
  frontend:
    mode: hybrid
    reverse: 
    - 3000:8080
    workdir: ./frontend
    command: ["yarn start"]
```

> in the example shown we have indicated that we develop the `frontend` service locally using the `mode: hybrid` field.  The service running locally will execute the command defined in the `command` field in the directory indicated in `workdir`. In addition, the application will expose on port `3000` of the development container what it exposes on port `8080` of the local machine, making it possible for other components running on the cluster to access the service as if it were running remotely.

Reviewer 4: ```suggestion
In this example we have configured the `frontend` service to use `mode: hybrid`. This means the `frontend` service will exist on our local machine and not in the development container. 

This service will execute (locally) the command defined in the `command` field and in the directory specified in `workdir`. Additionally, anything exposed locally on port `8080` will be exposed on port `3000` of the development container. This makes it possible for other components running on the cluster to access the local service as if it were running remotely in the cluster.
```

@pchico83 mind reviewing this to make sure I haven't misunderstood something? 🙏

Reviewer 2: The second paragraph looks great, but in the first one I wouldn't say "`frontend` service will exist on our local machine", I would say something "okteto will redirect requests to the `frontend` service to your local machine.
And the second paragraph could say:
"Okteto will execute (locally) the command..."

Author: 768595236ccf3f4be66c9cb95b547cd7d0de78b2

---
title: Hybrid development
description: Hybrid development allows you to develop and run a service locally while the other components of your app run in the cloud.
sidebar_label: Hybrid development
id: hybrid-development
---

import Image from '@theme/Image';

> Okteto offers two different approaches to service development: `sync` or `hybrid` mode. This mode is configured in your Okteto manifest and affects the behavior of [`okteto up`](/reference/cli/#up).

Reviewer 2: @omnomagonz not in this PR, but I think this page should explain how dev containers work. And as part of it, explain how `sync` and `hybrid` modes work

> `Sync` is the default mode and changes made to the local service are synchronized with the remote development container deployed in the cluster. Alternatively, `hybrid` mode enables you to run the service in development locally while keeping the rest of the application components in the cloud.

Reviewer 2: ```suggestion
`Sync` is the default mode and code changes made to the local filesystem are synchronized to the remote development container deployed in the cluster. Alternatively, `hybrid` mode enables you to run the service in development locally while keeping the rest of the application components in the cloud.
```

### Deciding which mode to use

There are situations when running a service locally is faster than synchronizing code with a remote development container. For example: when doing frontend-only development or when working in an IDE-heavy context (e.g. Java/Spring) where the IDE may be pre-configured to run tests, start services, and connect debuggers. 

When using Okteto's `hybrid` mode you can maintain your local development experience for the specific service you're working on while still leveraging the benefits of hosting the rest of your application's components in the remote development container.

> > Using `hybrid` development mode breaks the replicability between operating systems and increases latency when accessing the rest of the services in the cluster. 

Reviewer 2: ```suggestion
> Using `hybrid` might increase latency when accessing the rest of the services in the cluster. 
```
I would remove the mention to applicability. This can be solved by running `okteto up` inside a local container

### How hybrid development works

> To allow communication to and from the cluster and your local machine an SSH server is deployed in the remote development container. You can expose the local service to the development container using the [`reverse`](/reference/okteto-manifest.mdx/#reverse-string-optional) configuration for the development environment. If you want to forward a service exposed in the cluster to your local machine you can use the [`forward`](/reference/okteto-manifest.mdx/#forward-string-optional) configuration.

Reviewer 2: ```suggestion
To allow communication to and from the cluster and your local machine an SSH server is deployed in the remote development container. You can expose the local service to the services running in the cluster using the [`reverse`](/reference/okteto-manifest.mdx/#reverse-string-optional) configuration for the development environment. If you want to forward a service running in the cluster to your local machine you can use the [`forward`](/reference/okteto-manifest.mdx/#forward-string-optional) configuration.
```

Author: 9ed9c2367e2fee70d1c8a8e1a553175cea6459fd

<p align="center">
  <Image
    src={require("@site/static/img/hybrid-development-arch.png").default}
    alt="Hybrid development architecture"
    width="850"
  />
</p>

> When using `hybrid` mode the process running locally is designed to behave similarly as if it were running in `sync` mode. To achieve this the following environment variables are added to the local process running the service in `hybrid` mode. This list is ordered from highest to lowest prevalence:

Reviewer 2: ```suggestion
When using `hybrid` mode the process running locally is designed to run with the same environment as the original service running in the cluster. To achieve this, the following environment variables are added to the local process running the service in `hybrid` mode. This list is ordered from highest to lowest prevalence:
```

Author: 359692e888a6f51dc3f0bd5026832fbd7c780f10

> - Environment variables declared in the `dev` section of the manifest

Reviewer 2: @AdrianPedriza can we add a link to https://www.okteto.com/docs/reference/manifest/#environment-string-optional?

Author: ae356f614884e5efb1be8236b4369c11f6af7f6b

> - Environment variables of the remote development container

Reviewer 2: ```suggestion
- Environment variables defined in the original container spec (add link to container envars k8s docs)
```

Author: c254fd0cc7e0892e68bede41146296585d0f0e8e

> - Environment variables of the configmap associated with the dev environment

Reviewer 2: ```suggestion
- Environment variables coming from configmaps defined in the original container spec (add link to k8s docs)```

Reviewer 2: @AdrianPedriza I think you missed this comment

Author: oh yes I overlooked it. What k8s docs would be appropriate to link here?

Reviewer 2: this is a good link:
https://kubernetes.io/es/docs/concepts/configuration/configmap/#configmaps-y-pods

Author: great 5bc16e5b1befd0fa22206247f20ecf5a480dc0a9

> - User-defined variables

Reviewer 2: ```suggestion
- Environment variables coming from secrets defined in the original container spec (add link to k8s docs)
```

Author: what link should we use here?

Reviewer 2: https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/#define-container-environment-variables-using-secret-data

Author: c254fd0cc7e0892e68bede41146296585d0f0e8e

> - Environment variables declared in the image of the remote development container

Reviewer 2: ```suggestion
- Environment variables declared in the image of the original container (k8s docs link to the field `image`)
```

Author: a link to https://kubernetes.io/docs/concepts/containers/images/ is enough?

Author: c254fd0cc7e0892e68bede41146296585d0f0e8e

### How to develop in hybrid mode

> To enable `hybrid` mode you only need to configure the necessary ports to expose the local service to the remote development container. This allows the components running in the cluster to communicate with the local service as if it was running directly in the cluster. 

Reviewer 2: ```suggestion
To enable `hybrid` mode you only need to configure the necessary ports to expose the local service to the cluster network. This allows the components running in the cluster to communicate with the local service as if it was running directly in the cluster. 
```

Author: 8508a5d2898cebc9c26ac3b52202ae30c5d1cd54

To do this you can make use of the [reverse](reference/manifest/#reverse-string-optional) field in your Okteto manifest, which will expose the configured port of your local machine to the desired port on the development container.

```yaml
dev:
  frontend:
    mode: hybrid
    reverse: 
    - 3000:8080
    workdir: ./frontend
    command: ["yarn start"]
```

> In this example we have configured the `frontend` service to use `mode: hybrid`. This means Okteto will redirect requests to the frontend service to your local machine and not in the development container.

Reviewer 2: ```suggestion
In this example we have configured the `frontend` service to use `mode: hybrid`. This means Okteto will redirect requests to the frontend service in the cluster to your local machine.
```

Author: 8508a5d2898cebc9c26ac3b52202ae30c5d1cd54

> Okteto will execute (locally) the command defined in the `command` field and in the directory specified in `workdir`. Additionally, anything exposed locally on port `8080` will be exposed on port `3000` of the development container. This makes it possible for other components running on the cluster to access the local service as if it were running remotely in the cluster.

Reviewer 2: `Okteto will execute (locally) the command defined in the `command` field and in the directory specified in `workdir` -> I would mention here that we also autoinject environment variables to simulate the same environment than the original container. With a link to where this is documented

Author: a52b4aea3ea4ea89f11c21d64b43d6526b200670

import Image from '@theme/Image';

> Okteto offers two different approaches to service development: sync ot hybrid mode. The command affected by this is the [`okteto up`](/reference/cli/#up) command being configurable at manifest level.

Reviewer 4: ```suggestion
Okteto offers two different approaches to service development: `sync` or `hybrid` mode. This mode is configured in your Okteto manifest and affects the behavior of [`okteto up`](/reference/cli/#up). 
```

> The default mode used is the *sync* mode in which changes made to the local service are synchronized with the remote development container deployed in the cluster. On the other hand, the *hybrid* mode allows to run the development service locally and the rest of the application components in the cloud.

Reviewer 4: ```suggestion
`Sync` is the default mode and changes made to the local service are synchronized with the remote development container deployed in the cluster. Alternatively, `hybrid` mode enables you to run the service in development locally while keeping the rest of the application components in the cloud.
```

Author: 1d810a59e441c6d926b8b04c03046b45b82df0d2

### How hybrid development works?

> Usually, okteto allows you to develop a service locally and run it in a production-like cloud environment synchronizing the local changes with the development container running in the cloud (sync mode) but okteto also allows you to use another approach in which the service to be developed runs locally and the rest of the application components run in the cloud (hybrid mode).

Reviewer 4: ```suggestion
By default Okteto enables you to edit a service's code locally and have it run in a production-like cloud environment. This is achieved using `syncthing` where local changes are synchronized with the development container running in the cloud.

Okteto also allows you to use another mode called `hybrid` in which the service to be developed runs locally and the rest of the application components run in the cloud.
```

I'm wondering if we could just omit the first paragraph especially if it's mentioned elsewhere. It seems a bit odd to have a section header called "How does hybrid mode work" and then spent the first half talking about sync mode 🤔 

any opinions?

Author: agree

Author: check the preview after the changes in that section. Are you OK with it?

> Using *hybrid* mode a remote container is also raised because when a component of the cluster needs to consume the service being developed locally, the `reverse` is used to forward to the desired port of the development container what is exposed locally. In this way, for the rest of the cluster components the *hybrid* or *sync* mode is indifferent.

Reviewer: Not sure if this paragraph is clear to me. Let's try to define it in a way that explains what is shown in the diagram.

Author: ok! take a look at this paragraph and let me know if it is ok https://github.com/okteto/docs/pull/557/files#diff-4704534dddb0360cc8e8c8559f984492563b94235123bf13714fd4867ad9933aR23
